compiler/parser/RdrHsSyn.hs

```diff
@@ -2810,21 +2811,26 @@ mkTemplateInstanceDecl sharedBinds templateName conName ValidTemplate{..} =
     this = asPatRecWild "this" conName
     cid = mkVarPat $ mkVarOcc "cid"
 
-mkInterfaceMethodDecl :: LHsType GhcPs -> LHsType GhcPs -> Located RdrName -> LHsType GhcPs -> [LHsDecl GhcPs]
-mkInterfaceMethodDecl ifaceTy classTy methodName methodType =
-  [ hasMethodInstance
-  , topLevelFuncSig
+mkInterfaceMethodDecl :: LHsType GhcPs -> LHsType GhcPs -> Located RdrName -> LHsType GhcPs -> Maybe LHsDocString -> [LHsDecl GhcPs]
+mkInterfaceMethodDecl ifaceTy classTy methodName methodType mbDocString =
+  hasMethodInstance :
+  maybeToList mbDocDecl ++
+  [ topLevelFuncSig
   , topLevelFuncDef
   ]
   where
-    hasMethodInstance = instDecl $ classInstDecl instanceType emptyBag
+    mbDocDecl = fmap (\(L l docStr) -> L l (DocD (noExt) (DocCommentPrev docStr))) mbDocString
+    hasMethodInstance = addLoc loc $ instDecl $ classInstDecl instanceType emptyBag
       where
+        addLoc loc = L loc . unLoc
```

Comment:
I don’t quite understand this. What is the original location and why do we need to patch it up?

Reply:
The original location is `noLoc`. We need to have the location on the instance such that we can correlate parsed instance declarations together with their docnext/docprev and `ClsInst` from the typechecked module.

Reply:
why is the original location `noLoc`? Can we not set the correct location when parsing?

Reply:
the method instance is generated via `instDecl` which sets the location to `noLoc`. One could think about adding a location argument to `instDecl` but that would mean changing a lot of calls.

Reply:
alright let’s leave it for now but maybe worth tryting to do that later.